main/lsp/ExtractVariable.cc

```diff
@@ -13,66 +13,89 @@ void logDebugInfo(const std::shared_ptr<spdlog::logger> logger, const core::Glob
         if (classDef->rhs.size() == 0) {
             ENFORCE(false);
         } else {
             for (auto &stat : classDef->rhs) {
                 if (stat.loc().contains(target)) {
-                    whereToInsert = stat.loc();
+                    whereToInsert = &stat;
                     break;
                 }
             }
         }
     } else if (auto block = ast::cast_tree<ast::Block>(scope)) {
-        whereToInsert = block->body.loc();
+        whereToInsert = &block->body;
     } else if (auto methodDef = ast::cast_tree<ast::MethodDef>(scope)) {
-        whereToInsert = methodDef->rhs.loc();
+        whereToInsert = &methodDef->rhs;
     } else if (auto if_ = ast::cast_tree<ast::If>(scope)) {
         if (if_->thenp.loc().exists() && if_->thenp.loc().contains(target)) {
-            whereToInsert = if_->thenp.loc();
+            whereToInsert = &if_->thenp;
         } else if (if_->elsep.loc().exists() && if_->elsep.loc().contains(target)) {
-            whereToInsert = if_->elsep.loc();
+            whereToInsert = &if_->elsep;
         } else {
             ENFORCE(!(if_->cond.loc().exists() && if_->cond.loc().contains(target)),
                     "shouldn't be inserting into the cond of an if");
+            ENFORCE(false, "none of the if sub-parts contain the target");
         }
     } else if (auto rescue = ast::cast_tree<ast::Rescue>(scope)) {
         if (rescue->body.loc().exists() && rescue->body.loc().contains(target)) {
-            whereToInsert = rescue->body.loc();
+            whereToInsert = &rescue->body;
         } else if (rescue->else_.loc().exists() && rescue->else_.loc().contains(target)) {
-            whereToInsert = rescue->else_.loc();
+            whereToInsert = &rescue->else_;
         } else if (rescue->ensure.loc().exists() && rescue->ensure.loc().contains(target)) {
-            whereToInsert = rescue->ensure.loc();
+            whereToInsert = &rescue->ensure;
         } else {
             ENFORCE(false)
         }
     } else if (auto rescueCase = ast::cast_tree<ast::RescueCase>(scope)) {
-        whereToInsert = rescueCase->body.loc();
+        whereToInsert = &rescueCase->body;
     } else if (auto while_ = ast::cast_tree<ast::While>(scope)) {
         if (while_->body.loc().exists() && while_->body.loc().contains(target)) {
-            whereToInsert = while_->body.loc();
+            whereToInsert = &while_->body;
         } else {
             ENFORCE(!(while_->cond.loc().exists() && while_->cond.loc().contains(target)),
                     "shouldn't be inserting into the cond of a while");
+            ENFORCE(false, "none of the while sub-parts contain the target");
         }
     } else {
         ENFORCE(false);
     }
-    ENFORCE(whereToInsert.exists());
-    return whereToInsert;
+    ENFORCE(whereToInsert);
+    if (auto caseBodyLoc = detectCase(whereToInsert, target)) {
+        return *caseBodyLoc;
+    }
+    return whereToInsert->loc();
```

Comment:
As written, wouldn't this `whereToInsert->loc()` crash if we didn't find something?  We should be handling that gracefully rather than relying on `ENFORCE`s.

(We'll have the same problem with the `nullptr` change suggested above, it just becomes a little more obvious because there's no loc to associated with `nullptr`.)